Add USBCDC_ECM class

juhaylinen opened the pull request:
### Description
Add initial Ethernet over USB communication support using
the ECM (Ethernet Control Model) subclass of USB CDC.

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [x] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@c1728p9

juhaylinen committed:
Add USBCDC_ECM class

Add initial Ethernet over USB communication support using
the ECM (Ethernet Control Model) subclass of USB CDC.

ciarmcom commented:
@juhaylinen, thank you for your changes.
@c1728p9 @ARMmbed/mbed-os-maintainers please review.

juhaylinen commented:
FYI @mikter

c1728p9 commented on usb/device/USBCDC_ECM/USBCDC_ECM.cpp:
```diff
@@ -0,0 +1,496 @@
+
+    _bulk_buf_size = read_finish(_bulk_out);
+
+    read_start(_bulk_out, _bulk_buf, MAX_PACKET_SIZE_BULK);
```
Without being able to receive data the functionality of this device is quite limited.

c1728p9 commented on usb/device/USBCDC_ECM/USBCDC_ECM.cpp:
```diff
@@ -0,0 +1,496 @@
+    if (setup->bmRequestType.Type == CLASS_TYPE) {
+        //printf("In USBCallback_request: CLASS specific Request: %02x\n", setup->bRequest);
+        switch (setup->bRequest) {
+            case SET_ETHERNET_MULTICAST_FILTERS:
```
Should these be forwarded to the user of this call as a callback? Alternatively, if this class took in a an `EMAC` phy then you could call the appropriate functions. @kjbracey-arm what are your thoughts on this?

c1728p9 commented on usb/device/USBCDC_ECM/USBCDC_ECM.h:
```diff
@@ -0,0 +1,209 @@
+    void _bulk_in_callback();
+    void _bulk_out_callback();
+    bool _notify_network_connection(uint8_t value);
+    bool _notify_connection_speed_change(uint32_t up, uint32_t down);
```
`_notify_network_connection` and `_notify_connection_speed_change` are private and I don't see them being used anywhere in this class. Is this dead code?

c1728p9 commented on usb/device/USBCDC_ECM/USBCDC_ECM.cpp:
```diff
@@ -0,0 +1,496 @@
+
+const uint8_t *USBCDC_ECM::string_iconfiguration_desc()
+{
+    static uint8_t string_imac_addr[26] = {0};
```
This array should probably be a member of the class rather than static.

c1728p9 commented on usb/device/USBCDC_ECM/USBCDC_ECM.cpp:
```diff
@@ -0,0 +1,496 @@
+        return NULL;
+    }
+
+    static const uint8_t configDescriptor[] = {
```
This array should probably be a member of the class rather than static. Also the local static initialization will cause mutexes to be used in interrupt context.

c1728p9 reviewed (changes requested):
If read and write are implemented can linux then use this class to provide an internet connection over USB?

juhaylinen committed:
Fix USBCDC_ECM descriptors

Replace static arrays used for iMAC and config descriptor with class members
Minor coding style fixes

cmonr commented:
@c1728p9 Mind taking another look? New commit looks like it resolved old comments.

juhaylinen committed:
USBCDC_ECM: send connection notifications to the host

Fix build issues after rebase

juhaylinen committed:
USBCDC_ECM: Add receive functionality

Handle SET_ETHERNET_PACKET_FILTER request

juhaylinen commented:
@c1728p9 I have addressed the changes requested. Please review.

juhaylinen committed:
Fix mutex issue

bulislaw reviewed (approved):
Good arguments were raised during offline discussions. Specifically:
* Feature branch maintenance is expensive
* By pushing it to master we will add value
* Mbed OS promise is around development on 3 OSs, not feature support
* Feature can be used on OSs with standard ECM support, proprietary drivers can be explored when Windows support will be required
